src/hal/src/device.rs

```diff
@@ -411,11 +455,68 @@ pub trait Device<B: Backend> {
     fn create_fence(&self, signaled: bool) -> B::Fence;
 
     ///
-    fn reset_fences(&self, &[&B::Fence]);
+    fn reset_fence(&self, fence: &B::Fence) {
+        self.reset_fences(Some(fence));
+    }
+
+    ///
+    fn reset_fences<I>(&self, fences: I)
+    where
+        I: IntoIterator,
+        I::Item: Borrow<B::Fence>,
+    {
+        for fence in fences {
+            self.reset_fence(fence.borrow());
+        }
+    }
+
+    /// Blocks until the given fence is signaled.
+    /// Returns true if the fence was signaled before the timeout.
+    fn wait_for_fence(&self, fence: &B::Fence, timeout_ms: u32) -> bool {
+        self.wait_for_fences(Some(fence), WaitFor::All, timeout_ms)
+    }
 
     /// Blocks until all or one of the given fences are signaled.
     /// Returns true if fences were signaled before the timeout.
-    fn wait_for_fences(&self, &[&B::Fence], WaitFor, timeout_ms: u32) -> bool;
+    fn wait_for_fences<I>(&self, fences: I, wait: WaitFor, timeout_ms: u32) -> bool
+    where
+        I: IntoIterator,
+        I::Item: Borrow<B::Fence>,
+    {
+        use std::{thread, time};
+        let start = time::Instant::now();
+        fn as_ms(duration: time::Duration) -> u32 {
+            duration.as_secs() as u32 * 1000 + duration.subsec_nanos() / 1_000_000
+        }
+        match wait {
+            WaitFor::All => {
+                for fence in fences {
+                    let elapsed_ms = as_ms(start.elapsed());
+                    if elapsed_ms > timeout_ms {
+                        return false;
+                    }
+                    if !self.wait_for_fence(fence.borrow(), timeout_ms - elapsed_ms) {
+                        return false;
+                    }
+                }
+                true
+            },
+            WaitFor::Any => {
+                let fences: Vec<_> = fences.into_iter().collect();
+                loop {
+                    for fence in &fences {
+                        if self.wait_for_fence(fence.borrow(), 0) {
+                            return true;
+                        }
+                    }
+                    if as_ms(start.elapsed()) >= timeout_ms {
+                        return false;
+                    }
+                    thread::sleep(time::Duration::from_millis(1));
```

Comment:
I don't think we need issue a thread switch here, drivers will most likely do it internally anyways.